packages/controller-utils/src/util.ts

```diff
@@ -16,6 +16,19 @@ import { MAX_SAFE_CHAIN_ID } from './constants';
 
 const TIMEOUT_ERROR = new Error('timeout');
 
+const PROTOTYPE_POLLUTION_BLOCKLIST = ['__proto__', 'constructor'] as const;
+
+/**
+ * Checks whether a dynamic string used as an object property key
+ * could be used in a prototype pollution attack.
+ *
+ * @param key - The dynamic key to check for safety.
+ * @returns Whether the given dyanmic key is safe to use.
+ */
+export function isSafeDynamicKey(key: string): boolean {
+  return PROTOTYPE_POLLUTION_BLOCKLIST.every((badInput) => key !== badInput);
+}
```

Comment:
A few things:

- When people talk about a prototype pollution attack, do people call the offending items "keys" or "properties"? Some googling reveals that people use the latter, so to prevent confusion, do we want to use "property"?

- I'm confused by the use of "dynamic string" here. I know that prototype pollution refers to `__proto__` or `constructor` getting set to something that changes the behavior of the object or even all objects, and either of these property names could come from a variable and be set on the object dynamically. But that doesn't seem to be what "dynamic string" is communicating. I tried googling this term and didn't find much, so I feel like would confuse other people too.

- For a prototype pollution alert on GitHub's code scanning tool, if you click on "Show more", the last example excludes `prototype` as well. It's also included in the example on this CWE: https://cwe.mitre.org/data/definitions/915.html. So, should `prototype` be added to the blocklist too?

- Finally, I'm curious if the use of `includes` may be more readable (at the expense of using a type assertion)?

Putting this together, what about:

``` suggestion
const PROTOTYPE_POLLUTION_BLOCKLIST = ['__proto__', 'constructor', 'prototype'] as const;

/**
 * Checks whether the given property name could be used in a
 * [prototype pollution attack](https://learn.snyk.io/lesson/prototype-pollution/).
 *
 * @param propertyName - The property name to check.
 * @returns Whether the given property name is safe to set.
 */
export function isSafeProperty(propertyName: string): boolean {
  const blocklist = PROTOTYPE_POLLUTION_BLOCKLIST as string[];
  return !blocklist.includes(propertyName)
}
```

Reply:
In JavaScript at least, [a "property" is a key-value pair](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#properties). The measure we're implementing in this PR is summarized as ["sanitizing property keys"](https://portswigger.net/web-security/prototype-pollution/preventing#sanitizing-property-keys) by the makers of burpsuite. "(object) property key/name," or just "key" seem like the most rigorous terms to choose from.

The "dynamic string" was an error, especially since it's also used for string interpolation or dynamic resizing. I intended to refer to "dynamic (object/property) key" or "computed property name/key." The latter is the official term used by the [ES6 standard](https://262.ecma-international.org/6.0/#sec-object-initializer) and [eslint](https://eslint.org/docs/latest/rules/no-useless-computed-key), and would be the most unambiguous term we could use, but "dynamic key" is more widely used, and might be a bit easier to understand for non-experts in JavaScript.

Adding `prototype` to the blocklist sounds good to me! I'm not sure about the merits of switching to `includes` at the cost of a type assertion, especially if the internal logic of this one-liner function is intended to be encapsulated. I feel like `every` and `some` are intuitive and powerful tools we shouldn't shy away from using. Maybe negating the `every` expression might make it more readable?

```suggestion
const PROTOTYPE_POLLUTION_BLOCKLIST = ['__proto__', 'constructor', 'prototype'] as const;

/**
 * Checks whether a dynamic property key could be used in
 * a [prototype pollution attack](https://portswigger.net/web-security/prototype-pollution).
 *
 * @param key - The dynamic key to validate.
 * @returns Whether the given dynamic key is safe to use.
 */
export function isSafeDynamicKey(key: string): boolean {
  return !PROTOTYPE_POLLUTION_BLOCKLIST.some((blockedKey) => key === blockedKey);
}
```

Reply:
Okay, these changes make sense to me. The link you found also explains the problem more clearly.